File: Easy Solutions/C/MiddleOfLinkedList.c
//Runtime: 0 ms, beats 100.0% of C online submissions
//Memory Usage: 5.9 MB beats 58.98% of C online submissions

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


struct ListNode* middleNode(struct ListNode* head){
    //hare and turtle
    struct ListNode* fast = head;
    struct ListNode* slow = head;
    
    while (fast != NULL && fast->next != NULL) {
        fast=fast->next->next;
        slow=slow->next;
    }
    
    return slow;
}

File: Easy Solutions/Python/MiddleOfLinkedList.py
#Runtime: 28 ms, beats 85.38% of Python3 online submissions
#Memory Usage: 14.3 MB

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        #hare and turtle
        slow = fast = head

        while (fast != None and fast.next != None):
            slow = slow.next
            fast = fast.next.next

        return slow